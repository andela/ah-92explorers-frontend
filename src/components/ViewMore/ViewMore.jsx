import React, { Component, Fragment } from 'react';
import { connect } from 'react-redux';
import {
  ListGroup,
  ListGroupItem,
  Modal, ModalBody,
  ModalFooter,
  Button,
  ModalHeader,
} from 'reactstrap';
import NavBar from '../Layout/navBar';
import { getNotifications, readNotification } from '../../redux/actions/actionCreators';

export class ViewMore extends Component {
  state = {
    modal: false,
    notificationBody: '',
  }

  componentDidMount() {
    this.props.getNotifications();
  }

  toggle = (e) => {
    this.setState(prevState => ({
      modal: !prevState.modal,
    }));
  }

  onclickNotif = (e) => {
    this.setState({
      notificationBody: e.target.getAttribute('data-value'),
    });
    this.toggle(e);
    this.props.readNotification(e.target.getAttribute('data-key'));
  }

  refresh = () => {
    const current = window.location.pathname;
    window.location.replace(current);
  }

  render() {
    const { notifications, read } = this.props;
    return (
      localStorage.getItem('authenticated') ? (
        <Fragment>
          <NavBar />
          <ListGroup className="viewmoreList">
            {
                notifications
                && notifications.notifications
                && notifications.notifications.allNotification
                && notifications.notifications.allNotification.map(notifs => (
                  <ListGroupItem style={{ backgroundColor: notifs.status === 'read' ? '#fff' : '#edf2fa', cursor: 'pointer' }} onClick={this.onclickNotif} key={`${notifs.id}`} data-value={`${notifs.message}`} data-key={`${notifs.id}`}>
                    {notifs.message}
                    {' '}
                    {notifs.status === 'read' ? <span className="read">{notifs.status}</span> : <span className="unread">{notifs.status}</span>}
                  </ListGroupItem>
                ))
              }
          </ListGroup>
          <Modal isOpen={this.state.modal} toggle={this.toggle} className={this.props.className}>
            <ModalHeader toggle={this.toggle}>Read Notification</ModalHeader>
            <ModalBody>
              { this.state.notificationBody }
            </ModalBody>
            <ModalFooter>
              <Button color="primary" onClick={this.refresh}>close</Button>
            </ModalFooter>
          </Modal>
        </Fragment>
      ) : window.location.replace('/login')
    );
  }
}

function mapStateToProps(state) {
  return {
    notifications: state.notifications,
  };
}


export default connect(mapStateToProps, { getNotifications, readNotification })(ViewMore);
