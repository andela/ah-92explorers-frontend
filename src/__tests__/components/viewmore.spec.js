import React from 'react';
import { shallow } from 'enzyme';
import { Provider } from 'react-redux';
import promiseMiddleware from 'redux-promise-middleware';
import thunk from 'redux-thunk';
import configureMockStore from 'redux-mock-store';
import { ViewMore } from '../../components/ViewMore/ViewMore';

const middlewares = [thunk, promiseMiddleware];
const mockStore = configureMockStore(middlewares);
describe('Renders CreateArticleComponent', () => {
    jest.spyOn(window.localStorage.__proto__, 'setItem');
    window.localStorage.__proto__.setItem = jest.fn();
    window.localStorage.setItem('authenticated', true);
    const props = {
      notifications: {
        notifications: {
            allNotification: []
        },
      },
      getNotifications: jest.fn(),
      readNotification: jest.fn(),
    }
    const wrapper = shallow(
        <ViewMore {...props}/>
    );
    it('should render ViewMore component', () => {
      expect(wrapper).toMatchSnapshot();
    });
    
    it('test update state onchange', () => {
      wrapper.instance().componentDidMount()
      expect(wrapper.instance().props.getNotifications).toBeCalled();
    });
  
    it('shoudl handle on comment', () => {
      const component = wrapper.instance();
      component.onclickNotif({ target: { getAttribute: jest.fn()}});
      component.refresh();
      expect(component).toBeDefined();
    });
});
