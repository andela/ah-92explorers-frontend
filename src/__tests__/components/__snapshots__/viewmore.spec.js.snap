// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Renders CreateArticleComponent should render ViewMore component 1`] = `ShallowWrapper {}`;
